### PR TITLE
[JUJU-754] Updates SetLocation functionality to work with mod.

### PR DESCRIPTION
**Pull request by tlm**
- Since the switch to go mod this package has been broken in terms of
  how it reports error locations with tracing. Before go mod it was a
  reasonable assumption that project's using this package would be built
  on the go path and this library could reasonable filter out the go path
  from file names when reporting error locations.

  Now that go mod is available and go programs can be built anywhere
  this package was failing to perform this filtering. This change now
  uses the runtime package to gather the package.file.function and line
  number for tracing.

  The resultant format is different to the previous implementation but
  now provides a much more go stack trace feel that works with go mod.

- Unit tests in this package were also assuming this very similar
  behaviour which no longer works. Have updated all tests to work
  regardless of the testing path.

- Removed a pre test meta compiler from the package that tried to
  establish the locations of error generation for asserting tracing
  values. This is now a much more explicit string that is built up
  alongside the test.

### CODE REVIEW

File: package_test.go [hunk 1]
```diff
@@ -5,91 +5,27 @@ package errors_test
...
-		}
-		result += line[0:i]
-		line = line[i+1:]
-		i = strings.Index(line, "$")
-		if i == -1 {
-			panic("no second $")
-		}
-		result += location(line[0:i]).String()
-		line = line[i+1:]
+// errorLocationValue provides the filename and line number for where this function
```
Comment (wallyworld) [inline package_test.go:18]: s/filename/function name

File: error.go [hunk 2]
```diff
@@ -25,9 +25,11 @@ type Err struct {
 	// previous holds the previous error in the error stack, if any.
 	previous error
 
-	// file and line hold the source code location where the error was
-	// created.
-	file string
```
Comment (wallyworld) [inline error.go:30]: Can we keep file as well? Yes, the absolute path will be used, but it's still useful right?

**Review by wallyworld — APPROVED** [review]
I like it, but wonder if we should keep file as well.

File: error.go (hunk 2 above)
Comment (tlm) [reply error.go:30]: Not really. In a development build on your local machine yeah it's useful but as soon as we cut a release people would get values that make no sense with respect to their current system. Also the package name is the path relative to the source tree and you know where your source tree lives.

**Pull request merged**